### PR TITLE
Linspace for tensors

### PR DESCRIPTION
**Pull request by hristo-vrigazov**
Issue [32093](https://github.com/tensorflow/tensorflow/issues/32093)

Adds the possibility to run linspace on tensors, matching `numpy`'s behavior since `1.16.0`.

- Implemented a version that does linspace on tensors along a given axis. The only change in the signature of the function is adding an optional parameter axis. Let me know if this is a problem and it should be done in a separate, new operation.
- All unit tests from before are still passing (without changing them). Added unit tests for nd arrays. Added unit tests for corner cases (num=1). Compared the results to the outputs of the corresponding calls to `np.linspace`
-  If `num > 1`, `start` and `stop` are guaranteed to be the first and the last elements of the result. The result is built by concatenating start, evenly spaced numbers and stop along the specified axis.
- Added test for the case when num is a tensor and not a Python int
- Negative axes are handled correctly
- Changed the doc string of the parameter `err` of the method `assertNDArrayNear`, which is (I believe) incorrect: it says that it is `The maximum absolute difference allowed.`, while the function uses [`np.linalg.norm`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.norm.html) without specifying the order of the norm, hence it defaults to order 2 for vectors and matrices, which is different, e.g:
```python
a = np.array([1., 2.])
b = np.array([3., 5.])
print(np.linalg.norm(a - b)) # returns 3.6055
print(np.max(np.abs(a - b))) # returns 3.
```

Few things that I am not sure about and probably the more experienced reviewers can help with:
- Is adding a new parameter with default value a problem for backward compatibility?
- Because the implementation is different from the previous one (a raw kernel) and consisting of other operations (like `repeat`, `tile`, `range`, etc.), would this be a problem for example when deserializing a graph from previous versions?
- Currently, if `num==0` a tensor with a shape with an added axis whose value is zero is returned (which matches Numpy's behavior), but this is different from the old 0D implementation.

### CODE REVIEW

**Comment by tomhennigan** [comment]
Thanks for the PR! I suspect I am not the right reviewer. @nfelt has worked on linspace in Jan this year, perhaps they will have more context.

**Comment by nfelt** [comment]
So - I actually don't have very much context on linspace since my work in January was just a small bug fix.  I can take a look but since core TF isn't my primary area it may take a little while for me to do due diligence in reviewing this (especially since the new code is pretty different from the old approach).

@yongtang I see you might have a little context from the originating feature request issue #32093 - would you be up for reviewing this?

**Comment by yongtang** [comment]
@nfelt Sure. I can help review the PR.

**Comment by yongtang** [comment]
In TensorFlow, APIs change from time to time though backward compatibility is always preserved. 

In the past the "backward compatibility" is typically preserved within the same language binding. For example, in C++ when APIs are updated they are in the form of `Gather` vs `GatherV2`. For python APIs, backward-compatibility are preserved by extending additional args.

I haven't see the change from one language binding to another in yet.

In this PR the binding is from C++ to python. This could potentially be desirable as it reduces the need to maintain C++ (harder to maintain than python).

In case switch from C++ to python is accepted, the C++ kernel of `Linspace` may need to continue to exist for backward-compatibility. The performance of python implementation on top of primitive ops vs C++ Linspace may also need to be profiled, so that there is not a big drop in performance. 

Adding API review team to see if switching from C++ implementation to python (for `linspace`) is accepted. /cc @tensorflow/api-owners

**Comment by hristo-vrigazov** [comment]
Thanks a lot for the comments so far! I just wanted to add that:

* Moving this to a separate python operation (e.g `linspace_nd` or something is similar) is always possible
* I have used the Numpy implementation as a reference - the idea is to build the range from `start` and `stop`, then repeat it so that it gets the same shape as `start`, and then multiply that by the step nd array (`delta`), and shift it so that it starts from `start`.

File: tensorflow/python/ops/math_ops.py [hunk 1]
```diff
@@ -99,9 +99,49 @@
...
+  if not isinstance(axis, int):
+    raise TypeError('Axis should be an integer. Received an object {0} of type {1}'.format(axis, type(axis)))
+  with ops.name_scope(name, 'linspace', [start_in, stop_in]):
+    start = ops.convert_to_tensor(start_in, name='start')
+    # stop must be convertible to the same dtype as start
+    stop = ops.convert_to_tensor(stop_in, name='stop', dtype=start.dtype)
+    expanded_start = array_ops.expand_dims(start, axis=axis)
+    num_int = array_ops.convert_to_int_tensor(num, name='num')
+    num = cast(num_int, dtype=start.dtype)
+    shape = expanded_start.get_shape()
```
Comment (alextp) [inline tensorflow/python/ops/math_ops.py]: This won't work if expand_start has a partially unknown shape (use tf.shape(expand_start) instead of expand_start.get_shape)

File: tensorflow/python/ops/math_ops.py [hunk 2]
```diff
@@ -99,9 +99,49 @@
...
+    axis = array_ops.get_positive_axis(axis, shape.ndims)
+    expanded_stop = array_ops.expand_dims(stop, axis=axis)
+    # to avoid having negative values in the range or zero division
+    # The result is sliced in the end so a correct result is returned for num == 1.
+    range_end = gen_math_ops.maximum(num - 1., 1.)
+    delta = (expanded_stop - expanded_start) / range_end
+    shape_range = range(shape.ndims)
+    axis_tiled = array_ops.fill(shape_range.get_shape(), axis)
```
Comment (alextp) [inline tensorflow/python/ops/math_ops.py]: Why not tf.ones_like(shape_range) * axis?

File: tensorflow/python/ops/math_ops.py [hunk 3]
```diff
@@ -99,9 +99,49 @@
 from tensorflow.python.util.tf_export import tf_export
 
 # Aliases for some automatically-generated names.
-linspace = gen_math_ops.lin_space
+#linspace = gen_math_ops.lin_space
 nextafter = gen_math_ops.next_after
 
+def linspace(start_in, stop_in, num, axis=0, name=None):
```
Comment (alextp) [inline tensorflow/python/ops/math_ops.py]: The new argument has to be the last to avoid breaking backward compatibility

**Review by alextp — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@hristo-vrigazov Could you please check reviewer comments and keep us posted. Thanks!

**Comment by hristo-vrigazov** [comment]
Yes, absolutely, I will check this soon. I was very busy these days.

File: tensorflow/python/ops/math_ops.py (hunk 2 above)
Comment (hristo-vrigazov) [reply tensorflow/python/ops/math_ops.py]: Your proposal is indeed more elegant. Moreover, I think that with the `tf.ones_like` I am going to be able to handle `None` shapes as well (similar to one of your previous comments). I have changed this below, let me know if it's still an issue.

File: tensorflow/python/ops/math_ops.py (hunk 1 above)
Comment (hristo-vrigazov) [reply tensorflow/python/ops/math_ops.py]: Thanks! I added tests for this case - and changed it to use the runtime shape. I don't think we can support fully unknown shape though, we need to know the number of dimensions when the graph is being created because otherwise the `range` later fails.

File: tensorflow/python/ops/math_ops.py (hunk 3 above)
Comment (hristo-vrigazov) [reply tensorflow/python/ops/math_ops.py]: I think this is done - let me know if there are issues.

File: tensorflow/python/ops/math_ops.py [hunk 4]
```diff
@@ -99,9 +99,51 @@
...
+    range_indices = array_ops.reshape(num_range, reshape_target)
+    tiled_range_indices = array_ops.tile(range_indices, repeats)
+    start_repeated = array_ops.repeat_with_axis(expanded_start, num_fill, axis)
+    delta_repeated = array_ops.repeat_with_axis(delta, num_fill, axis)
+    res = start_repeated + delta_repeated * tiled_range_indices
+    concatenated = array_ops.concat((expanded_start, res, expanded_stop), axis=axis)
+    begin = array_ops.zeros_like(shape)
+    num_slice = ones_like_shape_range * num_int
+    size = array_ops.where_v2(mask, num_slice, shape)
+    return array_ops.slice(concatenated, begin, size)
```
Comment (hristo-vrigazov) [inline tensorflow/python/ops/math_ops.py]: There is an edge case here: if we pass `num == -1`, then instead of throwing an exception, we return a concatenated `start` and `stop`. I will check on how this can be validated properly.

File: tensorflow/python/ops/math_ops.py [hunk 5]
```diff
@@ -99,9 +99,51 @@
 from tensorflow.python.util.tf_export import tf_export
 
 # Aliases for some automatically-generated names.
-linspace = gen_math_ops.lin_space
+#linspace = gen_math_ops.lin_space
```
Comment (alextp) [inline tensorflow/python/ops/math_ops.py]: Remove commented-out code

File: tensorflow/python/ops/math_ops.py [hunk 6]
```diff
@@ -99,9 +99,51 @@
...
 nextafter = gen_math_ops.next_after
 
+def linspace(start_in, stop_in, num, name=None, axis=0):
```
Comment (alextp) [inline tensorflow/python/ops/math_ops.py]: Add docstring here or TF will have no documentation for this function

File: tensorflow/python/ops/math_ops.py [hunk 7]
```diff
@@ -99,9 +99,51 @@
...
+  if not isinstance(axis, int):
+    raise TypeError('Axis should be an integer. Received an object {0} of type {1}'.format(axis, type(axis)))
```
Comment (alextp) [inline tensorflow/python/ops/math_ops.py]: No need to make your own validation, I think TF's default validation should catch things here.

File: tensorflow/python/ops/math_ops.py [hunk 8]
```diff
@@ -99,9 +99,51 @@
...
+  with ops.name_scope(name, 'linspace', [start_in, stop_in]):
+    start = ops.convert_to_tensor(start_in, name='start')
+    # stop must be convertible to the same dtype as start
+    stop = ops.convert_to_tensor(stop_in, name='stop', dtype=start.dtype)
+    expanded_start = array_ops.expand_dims(start, axis=axis)
+    num_int = array_ops.convert_to_int_tensor(num, name='num')
+    num = cast(num_int, dtype=start.dtype)
+    shape = array_ops.shape(expanded_start)
+    ndims = shape.get_shape()[0].value
```
Comment (alextp) [inline tensorflow/python/ops/math_ops.py]: The code needs to handle the case when this is None

File: tensorflow/python/ops/math_ops.py [hunk 9]
```diff
@@ -99,9 +99,51 @@
...
+    axis = array_ops.get_positive_axis(axis, ndims)
+    expanded_stop = array_ops.expand_dims(stop, axis=axis)
+    # to avoid having negative values in the range or zero division
+    # The result is sliced in the end so a correct result is returned for num == 1.
+    range_end = gen_math_ops.maximum(num - 1., 1.)
+    delta = (expanded_stop - expanded_start) / range_end
+    shape_range = range(ndims)
```
Comment (alextp) [inline tensorflow/python/ops/math_ops.py]: ...which can be done by using `ndims = tf.shape(shape)[0]` above and `tf.range(ndims)` here

Comment (alextp) [reply tensorflow/python/ops/math_ops.py]: can you also add a test for unknown ndims? Making start a placeholder should do it.

**Review by alextp — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/math_ops.py (hunk 7 above)
Comment (hristo-vrigazov) [reply tensorflow/python/ops/math_ops.py]: How would it be able to catch it, is there a default validation for all parameters named `axis` somewhere?

File: tensorflow/python/ops/math_ops.py (hunk 9 above)
Comment (hristo-vrigazov) [reply tensorflow/python/ops/math_ops.py]: `tf.range(None)` does not seem to be supported, at least on my version of the master, so that's why I did not add it (see my comment above about not supporting fully unknown shapes). Am I missing something here?

File: tensorflow/python/ops/math_ops.py (hunk 7 above)
Comment (alextp) [reply tensorflow/python/ops/math_ops.py]: No, but expand_dims has a parameter named axis and will do this validation for you

Comment (hristo-vrigazov) [reply tensorflow/python/ops/math_ops.py]: Aaah, I see. Thanks, will do

File: tensorflow/python/ops/math_ops.py (hunk 9 above)
Comment (hristo-vrigazov) [reply tensorflow/python/ops/math_ops.py]: Edit I see what you mean - to get the runtime value, this way we won't be able to validate the axis though, and it looks like in similar operations - like for example `repeat_with_axis` that is used below - unknown `ndims` don't seem to be supported, so I am not sure how to proceed here.

Comment (alextp) [reply tensorflow/python/ops/math_ops.py]: Can't you replace repeat_with_axis with broadcast_to?

Comment (hristo-vrigazov) [reply tensorflow/python/ops/math_ops.py]: I definitely can, thanks! I will update the PR as soon as I have a few hours to work on this

**Comment by hristo-vrigazov** [comment]
Sorry for the mess, I did a mistake rebasing... Closing this and will reopen a new PR
